voice/src/main/kotlin/VoiceConnection.kt

```diff
@@ -0,0 +1,146 @@
+package dev.kord.voice
+
+import dev.kord.common.annotation.KordVoice
+import dev.kord.common.entity.Snowflake
+import dev.kord.gateway.Gateway
+import dev.kord.gateway.UpdateVoiceStatus
+import dev.kord.voice.gateway.VoiceGateway
+import dev.kord.voice.gateway.VoiceGatewayConfiguration
+import dev.kord.voice.handlers.UdpLifeCycleHandler
+import dev.kord.voice.handlers.VoiceUpdateEventHandler
+import dev.kord.voice.udp.AudioFramePoller
+import dev.kord.voice.udp.AudioFramePollerConfiguration
+import dev.kord.voice.udp.AudioFramePollerConfigurationBuilder
+import kotlinx.coroutines.*
+import mu.KotlinLogging
+import kotlin.contracts.ExperimentalContracts
+import kotlin.contracts.InvocationKind
+import kotlin.contracts.contract
+import kotlin.coroutines.CoroutineContext
+import kotlin.time.TimeSource
+
+/**
+ * Data that represents a [VoiceConnection], these will never change during the life time of a [VoiceConnection].
+ *
+ * @param selfId the id of the bot connecting to a voice channel.
+ * @param guildId the id of the guild that the bot is connecting to.
+ * @param sessionId the id of the current voice session, given by Discord.
+ */
+data class VoiceConnectionData(
+    val selfId: Snowflake,
+    val guildId: Snowflake,
+    val sessionId: String
+)
+
+private val voiceConnectionLogger = KotlinLogging.logger { }
+
+/**
+ * A [VoiceConnection] is an adapter that forms the concept of a voice connection, or a connection between you and Discord voice servers.
+ *
+ * @param gateway the [Gateway] that handles events for the guild this [VoiceConnection] represents.
+ * @param voiceGateway the underlying [VoiceGateway] for this voice connection.
+ * @param data the data representing this [VoiceConnection].
+ * @param voiceGatewayConfiguration the configuration used for [voiceGateway].
+ * @param audioProvider a [AudioProvider] that will provide [AudioFrame] when required.
+ * @param frameInterceptorFactory a factory for [FrameInterceptor]s that is used whenever audio is ready to be sent. See [FrameInterceptor] and [DefaultFrameInterceptor].
+ * @param voiceDispatcher the dispatcher used for this voice connection.
+ */
+@KordVoice
+class VoiceConnection(
+    val gateway: Gateway,
+    val voiceGateway: VoiceGateway,
+    val data: VoiceConnectionData,
+    internal var voiceGatewayConfiguration: VoiceGatewayConfiguration,
+    var audioProvider: AudioProvider,
+    var frameInterceptorFactory: (FrameInterceptorContext) -> FrameInterceptor,
+    voiceDispatcher: CoroutineDispatcher
```

Comment:
why are these vars?
and why the config is internal?

Reply:
Guess, I rushed this constructor, regarding the vars. Besides, config is changed in the VoiceUpdateHandler, and didn't think the end user should be changing this for *any* reason. Should I change the config to be public?

Reply:
mhm, it does make sense for the user to create their own voice connection if they have the data without the channel itself

Reply:
Well, if internal or not, the user can still create their `VoiceConnection` and provide their own initial voice gateway configuration, if they desire to do so. Though once it set, I don't see why a user would be changing it during the lifetime of a `VoiceConnection`, as that is done so properly by the `VoiceUpdateEventHandler`. Though I'll change it if you think it is best.

Reply:
example?

Reply:
Example of...? The initial gateway configuration is built in VoiceConnectionBuilder, Gateway#updateVoiceState. Any user could do this logic themselves and pass it into the VoiceConnection. Maybe we should make a public, read-only variable for the config, and an internal var that is changed by the VoiceUpdateEventHandler?? This seems reasonable to me...

Reply:
No, there is no need to create a default config variable, the user will edit the parts they care about and  also this val object if it's immutable the user will need to create a new copy any how